### DISCUSSION TITLE
Why do error occur when using ui.notify.

### DISCUSSION THREAD

**Discussion by akai-1024** [zauberzeug/nicegui#1006] [Q&A] [answered]
### Question

![ui.notify error.](https://raw.githubusercontent.com/akai-1024/self/main/error.png)

The code-usage scenario is that I built an event engine, and when an event was triggered, the call executed the function, but it made such an error.

The code is correct when I simply use it in a py file.
``` python
from const import ui


def handler_alert():
    a = ui.audio('https://cdn.pixabay.com/download/audio/2022/02/22/audio_d1718ab41b.mp3', controls=False, autoplay=True)
    a.on('ended', lambda _: ui.notify('Audio playback completed', type='negative'))


ui.button("Alert", on_click=handler_alert)
ui.run(native=True)
```

**Comment by falkoschindler** [comment]
Hi @akai-1024! It's hard to tell what's going on without seeing the code that is actually producing the error. Can you, please, provide a [minimal reproducible example](https://en.wikipedia.org/wiki/Minimal_reproducible_example)? Thanks!

**Reply by akai-1024** [reply]
Ok, this is the minimal example.
Probably because I introduce another asynchronous library AsyncIOScheduler in main process. Hope you can help me to look at this mistake.

Tips: When i use ui.timer to start the timed task, it's running success.

```python
# -*- coding:utf-8 -*-
# ========================================================================
# @File    : error_eventEngine.py
# @Date    : 2023-06-10
# @Author  : akai
# @Subject : ""
# ========================================================================
import asyncio
import inspect
import aiohttp
from collections import defaultdict
import pytz
from nicegui import ui
from apscheduler.schedulers.asyncio import AsyncIOScheduler


# ========================================================================
class EventManager:

    def __init__(self):
        self.__handler_dict = defaultdict(list)

    def register(self, _type, _func):
        self.__handler_dict[_type].append(_func)

    def emit(self, _type, _msg):
        handler_list = self.__handler_dict.get(_type)
        for handler in handler_list:
            if inspect.iscoroutinefunction(handler):
                asyncio.get_running_loop().create_task(handler(_msg))
            else:
                handler(_msg)


eventEngine = EventManager()


# ========================================================================
async def handler_mock_request():
    async with aiohttp.ClientSession() as session:
        res = await session.get(url="https://www.baidu.com")

        _type = "Response"
        _msg = "Request baidu is success."
        eventEngine.emit(_type, _msg)


# ========================================================================
class SchedulerManager:

    tz_shanghai = pytz.timezone("Asia/Shanghai")

    def __init__(self):
        self.scheduler = AsyncIOScheduler(timezone=self.tz_shanghai)
        self.scheduler.add_job(handler_mock_request, trigger="interval", seconds=5)

    def start(self):
        if self.scheduler.running:
            return
        self.scheduler.start()
        print("scheduler started.")


# ========================================================================
def handler_executor(_msg):
    ui.notify(_msg, type="negative")
    # try:
    #     ui.notify(_msg, type="negative")
    # except Exception as e:
    #     print(e)


# Register event handler
eventEngine.register("Response", handler_executor)

# Start scheduler
SchedulerCenter = SchedulerManager()
ui.button("Start", on_click=SchedulerCenter.start)

# Start main
ui.run(native=True)

```

The erorr is :
![](https://github.com/akai-1024/self/blob/main/error2.png?raw=true)

**Reply by falkoschindler** [reply, marked as answer]
The problem is that when `handler_executor` is executed, it's not clear on which client the notification should be shown. Usually NiceGUI automatically uses the right client. But in this case - because the execution is triggered from a different event manager -, there is no client assigned to this function.

But you can set the client explicitly. Since you're using the [auto-index page](https://nicegui.io/documentation#auto-index_page) and no page decorators, you can simply refer to the auto-index client like this:
```py
from nicegui import globals

...

def handler_executor(_msg):
    with globals.index_client:
        ui.notify(_msg, type="negative")
```

By the way, NiceGUI's integrated API for background tasks might be a handy shortcut for `asyncio.get_running_loop().create_task`:

```py
from nicegui import background_tasks

background_tasks.create(handler(_msg))
```

**Reply by rodja** [reply]
> NiceGUI's integrated API for background tasks might be a handy shortcut for `asyncio.get_running_loop().create_task`

And our `background_tasks.create` makes also sure exceptions are printed to the console instead of silently getting lost...

**Reply by akai-1024** [reply]
Thank you very much for your answer and helping me solve the problem

**Reply by Liquidmasl** [reply]
Its been a while but i ran into the same issue, but it seams the globals import is not working anymore? globals does not seam to be in the nicegui index.
But I found Client and its auto_index property, so that might work

but i tried using your background_task, and it seams to block the ui thread, so not really running in the background, but i cant find any documentation on it. did i misunderstand its use?

**Reply by falkoschindler** [reply]
Yes, the `Client` class is the place to look for the "auto-index client".

The blocking background task sounds like a separate topic. If the function you're calling with `background_tasks` runs synchronously, it will block the UI. It is rather meant for async functions which you don't want to await. See [this FAQ](https://github.com/zauberzeug/nicegui/wiki/FAQs#why-is-my-long-running-function-blocking-ui-updates) for an alternative approach using `run.ui_bound` and `run.cpu_bound`. If the problem remains, feel free to open a new Q&A discussion including a minimal reproducible example, so we can better see what you're doing. Thanks!

**Comment by Liquidmasl** [comment]
Thank you for the quick response!
For now i am just sticking with a boring old process, but i will look into it. Thank you!
